core/shared/src/main/scala/scalaz/zio/Async.scala

```diff
@@ -13,28 +13,20 @@ import scalaz.zio.ExitResult.Cause
  * form `Throwable => IO[Nothing, Unit]`
  */
 sealed abstract class Async[+E, +A] extends Product with Serializable { self =>
-  def fold[E1, B](
+  final def fold[E1, B](
     f: A => ExitResult[E1, B],
     g: Cause[E] => ExitResult[E1, B]
   ): Async[E1, B] =
     self match {
-      case Async.Now(r)          => Async.Now(r.fold(f, g))
-      case Async.MaybeLater(c)   => Async.MaybeLater(c)
-      case Async.MaybeLaterIO(c) => Async.MaybeLaterIO(c)
+      case Async.Now(r)            => Async.Now(r.fold(f, g))
+      case x @ Async.MaybeLater(_) => x
```

Comment:
No need for the as-pattern here

Reply:
It's for performance reasons. If I reconstruct, then Scala will create a new `MaybeLater`. Because the type is `[Nothing, Nothing]`, however, I can reuse the old one.